### ISSUE TITLE
New scheme for Privacy State Token: Scrappy

### ISSUE THREAD

**Issue by akakou** [antifraudcg/proposals#21]
This proposal is making Scrappy(SeCure Rate Assuring Protocol with PrivacY), 
a new scheme with strong privacy for Privacy State Token..

Scrappy is a rate-limiter the same as Privacy Pass, but this restants to the timing correlation attack.
Also, Scrappy works as E2E (between users and web service) in the hot paths (i.e., sign and verify), 
so that the issuer does not need to receive high access from users.

**Slides:** https://speakerdeck.com/akakou/scrappy-and-view-of-applying-to-web
**Base paper:** https://www.ndss-symposium.org/ndss-paper/scrappy-secure-rate-assuring-protocol-with-privacy/

### Note

- Although Scrappy mainly assumes the user has a unique resource in the hardware device (in particular, the TPM), 
we can use other unique resources, the same as the Trust Token API.
- TODO

### Other Reference

Privacy State Token:
https://developers.google.com/privacy-sandbox/protections/private-state-tokens?hl=en

**Comment by akakou** [comment]
This is Abstracts of Scrappy

> Preventing abusive activities caused by adversaries accessing online services at a rate exceeding that expected by websites has become an ever-increasing problem. CAPTCHAs and SMS authentication are widely used to provide a solution by implementing rate limiting, although they are becoming less effective, and some are considered privacy-invasive. In light of this, many studies have proposed better rate-limiting systems that protect the privacy of legitimate users while blocking malicious actors. However, they suffer from one or more shortcomings: (1) assume trust in the underlying hardware and (2) are vulnerable to side-channel attacks.
Motivated by the aforementioned issues, this paper proposes Scrappy: SeCure Rate Assuring Protocol with PrivacY. Scrappy allows clients to generate unforgeable yet unlinkable rate-assuring proofs, which provides the server with cryptographic guarantees that the client is not misbehaving. We design Scrappy using a combination of DAA and hardware security devices. Scrappy is implemented over three types of devices, including one that can immediately be deployed in the real world. Our baseline evaluation shows that the end-to-end latency of Scrappy is minimal, taking only 0.32 seconds, and uses only 679 bytes of bandwidth when transferring necessary data. We also conduct an extensive security evaluation, showing that the rate-limiting capability of Scrappy is unaffected even if the hardware security device is compromised.

**Comment by akakou** [comment]
@dvorak42 

I would like to ask that is this repository a proper place to propose this?
Or is it better to propose it to [this repository](https://github.com/WICG/trust-token-api)?

**Comment by dvorak42** [comment]
Is the particular proposal to add support for this as a new token type for the Private State Token API (fka Trust Token)? If so, then the https://github.com/WICG/trust-token-api is probably a better place to discuss that. If you'd like to present this variant of a token to the AFCG, then here is probably right to have that discussion.

**Comment by akakou** [comment]
Thank you for your answers!

I think that I would like to propose Scrappy as one of the token types for Private State Token. 
This was because the use case and property are similar, and we can reuse the discussion in the past.

(Now I have found that Privacy State Token's interface is slightly different from Scrappy's required one. 
Therefore, if it is hard to fill the gap, maybe we will go back to AFCG.)

**Comment by akakou** [comment]
@dvorak42 
Is this reasonable? I would like to know your opinion if you are ok.

**Comment by dvorak42** [comment]
If you can open an issue on the other repo, we can discuss more there. I think some of the attributes of Scrappy might be interesting for PST, though I haven't had time to do an in-depth read of it yet. I'll point out the issue to other folks on the team to see if they have thoughts.

**Comment by akakou** [comment]
@dvorak42 

I re-read your comments and may misread `the other repo` you said.
Did you mean the repository for Private State Token or my new repository?

**Comment by dvorak42** [comment]
Yes, the Private State Tokens repository. Unfortunately with IETF happening this week and preparation for it, I haven't had a chance to review the issue in more detail yet.

**Comment by akakou** [comment]
@dvorak42 
Thank you for your answer. I see!

Regarding the review, please don't worry. 
There is no reason to hurry, and I hope you review this when you are free.

Have fun at the IETF meeting!

**Comment by akakou** [comment]
By the way, I could present it at the community group meeting if Scrappy is a little too complex.
(However, due to my low English listening skills, I hope to respond to text questions instead of voice conversations.)

**Comment by dvorak42** [comment]
This week's meeting is a bit full, but perhaps you'd be able to present it at one of the April AFCG meetings?

**Comment by akakou** [comment]
@dvorak42 
Good! How about the meeting on April 13th?

**Comment by dvorak42** [comment]
That should work. I'll add you to the agenda for that date. (note the time
change for AFCG meetings due to Daylights Saving Time)

**Comment by akakou** [comment]
Thank you!!
I'm looking forward to presenting it on that day.

**Comment by akakou** [comment]
We discussed in the meeting:

https://github.com/antifraudcg/meetings/blob/main/2024/04-12.md

**Comment by akakou** [comment]
Sorry, the link to the slide was incorrect, but I have now fixed it.

**Comment by akakou** [comment]
We might be able to integrate Scrappy with [this discussions](https://www.ietf.org/archive/id/draft-ladd-privacypass-bbs-01.html).

**Comment by akakou** [comment]
I closed this issue because it should be merge to #22.

For future reference, to briefly summarize this approach: it is almost the same as the rate-limiting approach discussed in #22. If I had to point out a difference, it comes down to the following trade-offs.

Pros:
- Keys can be securely stored using a TPM.
- The design is simple (it only changes the parameters given to an existing well-known anonymous authentication scheme, DAA).


Cons:
- Rather than “anonymity” per se, it focuses on preventing tracking. As a result, the goal is to ensure that users across different counters/periods cannot be linked, which means the service ends up being given a counter (see [detail](https://mailarchive.ietf.org/arch/msg/privacy-pass/Z9HInzRy09-Tv9QvAbureVaUzdM/)).
